target/linux/ipq806x/files-6.6/arch/arm/boot/dts/qcom/qcom-ipq8064-wpq864.dts

```diff
@@ -404,8 +404,99 @@
 	pinctrl-names = "default";
 };
 
-&flash {
-	compatible = "jedec,spi-nor";
+&gsbi5 {
+	status = "okay";
+	qcom,mode = <GSBI_PROT_SPI>;
+
+	spi@1a280000 {
+		status = "okay";
+
+		pinctrl-0 = <&spi_pins>;
+		pinctrl-names = "default";
+
+		cs-gpios = <&qcom_pinmux 20 GPIO_ACTIVE_HIGH>;
+
+		flash@0 {
+			compatible = "jedec,spi-nor";
+			spi-max-frequency = <50000000>;
+			reg = <0>;
+
+			partitions {
+				compatible = "fixed-partitions";
+				#address-cells = <1>;
+				#size-cells = <1>;
+
+				partition@0 {
+					label = "1:SBL1";
+					reg = <0x0 0x20000>;
+					read-only;
+				};
+
+				partition@20000 {
+					label = "1:MIBIB";
+					reg = <0x20000 0x20000>;
+					read-only;
+				};
+
+				partition@40000 {
+					label = "1:SBL2";
+					reg = <0x40000 0x40000>;
+					read-only;
+				};
+
+				partition@80000 {
+					label = "1:SBL3";
+					reg = <0x80000 0x80000>;
+					read-only;
+				};
+
+				partition@100000 {
+					label = "1:DDRCONFIG";
+					reg = <0x100000 0x10000>;
+					read-only;
+				};
+
+				partition@110000 {
+					label = "1:TZ";
+					reg = <0x110000 0x80000>;
+					read-only;
+				};
+
+				partition@190000 {
+					label = "1:RPM";
+					reg = <0x190000 0x80000>;
+					read-only;
+				};
+
+				partition@2100000 {
+					label = "1:APPSBL";
+					reg = <0x210000 0x80000>;
+					read-only;
+				};
+
+				partition@290000 {
+					label = "1:APPSBLENV";
```

Comment:
You might run into problems with two `APPSBLENV`:

https://github.com/openwrt/openwrt/blob/4f87a4d84f3d6d1625962413ea56cca1ce83db7c/package/boot/uboot-envtools/files/ipq806x#L14-L15

Reply:
I see. 
The first 3 parts of the fist matching line are process. That is 0:APPSBLENV
If the nand:0 flash is not available(corrupt,erased) then the copy 1:APPSBLENV
I from nand:1 can be used.
```
ubootenv_mtdinfo () {
	UBOOTENV_PART=$(cat /proc/mtd | grep APPSBLENV)
	mtd_dev=$(echo $UBOOTENV_PART | awk '{print $1}' | sed 's/:$//')
	mtd_size=$(echo $UBOOTENV_PART | awk '{print "0x"$2}')
	mtd_erase=$(echo $UBOOTENV_PART | awk '{print "0x"$3}')
```